Unpin pinned dependency versions

**Describe the bug**

Currently, the following dependencies are pinned in `environment.yml`:

 - `werkzeug <2.2  # >=2.2 slows down S3 tests (deps: moto->flask->werkzeug)`
 - `xarray >=2022.6,<2023.09.0 # because of issue https://github.com/dcs4cop/xcube/issues/897`

In the case of xarray, 61 unit tests fail if the current latest version (2024.2.0) is installed. Of course these pins can't be retained indefinitely, and both pinned packages are already significantly behind the current version. We should fix the incompatibilities and unpin them as soon as possible.

Summary of failing tests with 2024.2.0:

```text
FAILED test/cli/test_dump.py::DumpTest::test_dump_ds - AssertionError: '<xarray.Dataset> Size: 8MB\nDimensions:        (lon: 360, ...
FAILED test/cli/test_level.py::LevelDataTest::test_all_defaults - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_level.py::LevelDataTest::test_invalid_inputs - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_level.py::LevelDataTest::test_level_with_nc - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_level.py::LevelDataTest::test_level_with_zarr - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_level.py::LevelDataTest::test_level_with_zarr_agg_method - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_level.py::LevelDataTest::test_level_with_zarr_agg_methods - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_level.py::LevelDataTest::test_level_with_zarr_link - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_level.py::LevelDataTest::test_level_with_zarr_num_levels_max - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_level.py::LevelDataTest::test_with_existing_output - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_level.py::LevelDataTest::test_with_output - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_level.py::LevelDataTest::test_with_tile_size_and_num_levels - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_optimize.py::OptimizeDataTest::test_defaults - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_optimize.py::OptimizeDataTest::test_in_place - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_optimize.py::OptimizeDataTest::test_user_output - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_prune.py::PruneDataTest::test_delete_block_file - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_prune.py::PruneDataTest::test_dry_run - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/cli/test_prune.py::PruneDataTest::test_no_dry_run - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/mldataset/test_fs.py::FsMultiLevelDatasetTest::test_io_nl_4_ts_256 - xcube.core.mldataset.fs.FsMultiLevelDatasetError: Failed to write dataset t...
FAILED test/core/mldataset/test_fs.py::FsMultiLevelDatasetTest::test_io_nl_4_ts_256_agg - xcube.core.mldataset.fs.FsMultiLevelDatasetError: Failed to write dataset t...
FAILED test/core/store/fs/test_registry.py::FileFsDataStoresTest::test_dataset_levels - xcube.core.store.error.DataStoreError: Failed to write dataset /tmp/xcubeiv...
FAILED test/core/store/fs/test_registry.py::FileFsDataStoresTest::test_dataset_netcdf - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/store/fs/test_registry.py::FileFsDataStoresTest::test_dataset_zarr - xcube.core.store.error.DataStoreError: Failed to write dataset '/tmp/xcubew...
FAILED test/core/store/fs/test_registry.py::FileFsDataStoresTest::test_mldataset_levels - xcube.core.store.error.DataStoreError: Failed to write dataset /tmp/xcubeu0...
FAILED test/core/store/fs/test_registry.py::MemoryFsDataStoresTest::test_dataset_levels - xcube.core.store.error.DataStoreError: Failed to write dataset xcube/testin...
FAILED test/core/store/fs/test_registry.py::MemoryFsDataStoresTest::test_dataset_netcdf - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/store/fs/test_registry.py::MemoryFsDataStoresTest::test_dataset_zarr - xcube.core.store.error.DataStoreError: Failed to write dataset 'xcube/testi...
FAILED test/core/store/fs/test_registry.py::MemoryFsDataStoresTest::test_mldataset_levels - xcube.core.store.error.DataStoreError: Failed to write dataset xcube/testin...
FAILED test/core/store/fs/test_registry.py::S3FsDataStoresTest::test_dataset_levels - xcube.core.store.error.DataStoreError: Failed to write dataset xcube/testin...
FAILED test/core/store/fs/test_registry.py::S3FsDataStoresTest::test_dataset_netcdf - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/store/fs/test_registry.py::S3FsDataStoresTest::test_dataset_zarr - xcube.core.store.error.DataStoreError: Failed to write dataset 'xcube/testi...
FAILED test/core/store/fs/test_registry.py::S3FsDataStoresTest::test_mldataset_levels - xcube.core.store.error.DataStoreError: Failed to write dataset xcube/testin...
FAILED test/core/test_dump.py::DumpDatasetTest::test_dump_dataset - AssertionError: "<xarray.DataArray 'precipitation' (time: 5, lat: 180, lon:...
FAILED test/core/test_normalize.py::TestNormalize::test_normalize_zonal_lat_lon - ValueError: cannot re-index or align objects with conflicting indexes found...
FAILED test/core/test_optimize.py::OptimizeDatasetTest::test_failures - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_optimize.py::OptimizeDatasetTest::test_optimize_dataset - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_optimize.py::OptimizeDatasetTest::test_optimize_dataset_in_place - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_optimize.py::OptimizeDatasetTest::test_optimize_dataset_unchunk_coords_str - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_optimize.py::OptimizeDatasetTest::test_optimize_dataset_unchunk_coords_str_in_place - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_optimize.py::OptimizeDatasetTest::test_optimize_dataset_unchunk_coords_str_list - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_optimize.py::OptimizeDatasetTest::test_optimize_dataset_unchunk_coords_str_list_in_place - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_optimize.py::OptimizeDatasetTest::test_optimize_dataset_unchunk_coords_str_tuple - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_optimize.py::OptimizeDatasetTest::test_optimize_dataset_unchunk_coords_str_tuple_in_place - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_optimize.py::OptimizeDatasetTest::test_optimize_dataset_unchunk_coords_true - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_optimize.py::OptimizeDatasetTest::test_optimize_dataset_unchunk_coords_true_in_place - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_timeslice.py::TimeSliceTest::test_append_time_slice - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_timeslice.py::TimeSliceTest::test_find_time_slice - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_timeslice.py::TimeSliceTest::test_insert_time_slice - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_timeslice.py::TimeSliceTest::test_replace_time_slice - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_timeslice.py::TimeSliceTest::test_update_corrupt_cube - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_timeslice.py::ZarrStoreTest::test_local - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_unchunk.py::UnchunkDatasetTest::test_chunked - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_unchunk.py::UnchunkDatasetTest::test_dont_unchunk_if_unchunked - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_unchunk.py::UnchunkDatasetTest::test_unchunk_all - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_unchunk.py::UnchunkDatasetTest::test_unchunk_coord_var - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_unchunk.py::UnchunkDatasetTest::test_unchunk_coords_only - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_unchunk.py::UnchunkDatasetTest::test_unchunk_data_var - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_unchunk.py::UnchunkDatasetTest::test_unchunk_data_var_coords_only - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_unchunk.py::UnchunkDatasetTest::test_unchunk_invalid_path - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/test_unchunk.py::UnchunkDatasetTest::test_unchunk_invalid_vars - ValueError: When encoding chunked arrays of datetime values, both the units...
FAILED test/core/zarrstore/test_generic.py::CommonS3ZarrStoreTest::test_it - ValueError: When encoding chunked arrays of datetime values, both the units...
ERROR test/core/store/fs/test_registry.py::NewCubeDataTestMixin::test_open_packed - ValueError: When encoding chunked arrays of datetime values, both the units...
ERROR test/core/store/fs/test_registry.py::NewCubeDataTestMixin::test_open_unpacked - ValueError: When encoding chunked arrays of datetime values, both the units...
= 61 failed, 1356 passed, 39 skipped, 6357 warnings, 2 errors in 212.92s (0:03:32) =
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,12 +1,8 @@
 ## Changes in 1.4.1 (in development)
 
-### Enhancements
-
-### Fixes
-
-### Other changes
-
 * Minor updates to make xcube compatible with pandas 2 and Python 3.12. (#933)
+
+* Minor updates to make xcube compatible with xarray >=2023.9.0. (#897, #939)
 
 ## Changes in 1.4.0
 

--- a/environment.yml
+++ b/environment.yml
@@ -44,7 +44,7 @@ dependencies:
   - tornado >=6.0
   - urllib3 >=1.26
   - werkzeug
-  - xarray >=2022.6,<2023.09.0 # because of issue https://github.com/dcs4cop/xcube/issues/897
+  - xarray >=2022.6
   - zarr >=2.11
   # Required by Coiled
   # These are very likely transitive deps anyway

--- a/environment.yml
+++ b/environment.yml
@@ -43,7 +43,7 @@ dependencies:
   - shapely >=1.6
   - tornado >=6.0
   - urllib3 >=1.26
-  - werkzeug <2.2  # >=2.2 slows down S3 tests (deps: moto->flask->werkzeug)
+  - werkzeug
   - xarray >=2022.6,<2023.09.0 # because of issue https://github.com/dcs4cop/xcube/issues/897
   - zarr >=2.11
   # Required by Coiled

--- a/test/cli/test_dump.py
+++ b/test/cli/test_dump.py
@@ -9,73 +9,34 @@ class DumpTest(CliDataTest):
         with xr.set_options(display_width=80):
             result = self.invoke_cli(["dump", TEST_NC_FILE])
 
-        # xarray v0.16.0 changed line truncation behaviour (see
-        # https://github.com/pydata/xarray/issues/3759 ), so this test checks
-        # the output against two expected values to cover both pre- and
-        # post-change behaviour.
-        output_template = (
-            "<xarray.Dataset>\n"
-            "Dimensions:        (time: 5, lat: 180, lon: 360, bnds: 2)\n"
-            "Coordinates:\n"
-            "  * lon            (lon) float64 -179.5 -178.5 -177.5 ... 177.5 178.5 179.5\n"
-            "  * lat            (lat) float64 -89.5 -88.5 -87.5 -86.5 ... 86.5 87.5 88.5 89.5\n"
-            "  * time           (time) datetime64[ns] 2010-01-01T12:00:00 ... 2010-01-05T1%s\n"
-            "    lon_bnds       (lon, bnds) float64 ...\n"
-            "    lat_bnds       (lat, bnds) float64 ...\n"
-            "    time_bnds      (time, bnds) datetime64[ns] ...\n"
-            "Dimensions without coordinates: bnds\n"
-            "Data variables:\n"
-            "    precipitation  (time, lat, lon) float64 ...\n"
-            "    temperature    (time, lat, lon) float64 ...\n"
-            "    soil_moisture  (time, lat, lon) float64 ...\n"
-            "Attributes:\n"
-            "    Conventions:           CF-1.7\n"
-            "    title:                 Test Cube\n"
-            "    time_coverage_start:   2010-01-01T00:00:00\n"
-            "    time_coverage_end:     2010-01-06T00:00:00\n"
-            "    geospatial_lon_min:    -180.0\n"
-            "    geospatial_lon_max:    180.0\n"
-            "    geospatial_lon_units:  degrees_east\n"
-            "    geospatial_lat_min:    -90.0\n"
-            "    geospatial_lat_max:    90.0\n"
-            "    geospatial_lat_units:  degrees_north\n"
-        )
-        # xarray v2022.06.0 changed dimension order (see
-        # https://github.com/pydata/xarray/pull/6601 and
-        # https://github.com/pydata/xarray/issues/6600)
-        output_template_1 = (
-            "<xarray.Dataset>\n"
-            "Dimensions:        (lon: 360, lat: 180, time: 5, bnds: 2)\n"
-            "Coordinates:\n"
-            "  * lon            (lon) float64 -179.5 -178.5 -177.5 ... 177.5 178.5 179.5\n"
-            "  * lat            (lat) float64 -89.5 -88.5 -87.5 -86.5 ... 86.5 87.5 88.5 89.5\n"
-            "  * time           (time) datetime64[ns] 2010-01-01T12:00:00 ... 2010-01-05T1%s\n"
-            "    lon_bnds       (lon, bnds) float64 ...\n"
-            "    lat_bnds       (lat, bnds) float64 ...\n"
-            "    time_bnds      (time, bnds) datetime64[ns] ...\n"
-            "Dimensions without coordinates: bnds\n"
-            "Data variables:\n"
-            "    precipitation  (time, lat, lon) float64 ...\n"
-            "    temperature    (time, lat, lon) float64 ...\n"
-            "    soil_moisture  (time, lat, lon) float64 ...\n"
-            "Attributes:\n"
-            "    Conventions:           CF-1.7\n"
-            "    title:                 Test Cube\n"
-            "    time_coverage_start:   2010-01-01T00:00:00\n"
-            "    time_coverage_end:     2010-01-06T00:00:00\n"
-            "    geospatial_lon_min:    -180.0\n"
-            "    geospatial_lon_max:    180.0\n"
-            "    geospatial_lon_units:  degrees_east\n"
-            "    geospatial_lat_min:    -90.0\n"
-            "    geospatial_lat_max:    90.0\n"
-            "    geospatial_lat_units:  degrees_north\n"
-        )
+        # Use a regex to accommodate the differing output formats produced by
+        # xarray 2023.8.0 and 2024.2.0.
+        output_regex = r"""<xarray.Dataset>( Size: 8MB)?
+Dimensions:        \(lon: 360, lat: 180, time: 5, bnds: 2\)
+Coordinates:
+  \* lon            \(lon\) float64 (3kB )?-179\.5 -178\.5 -177\.5 \.\.\. 177\.5 178\.5 179\.5
+  \* lat            \(lat\) float64 (1kB )?-89\.5 -88\.5 -87\.5 -86\.5 \.\.\. (86\.5 )?87\.5 88\.5 89\.5
+  \* time           \(time\) datetime64\[ns\] (40B )?2010-01-01T12:00:00 \.\.\. 2010-01-(05T1)?\.\.\.
+    lon_bnds       \(lon, bnds\) float64 (6kB )?\.\.\.
+    lat_bnds       \(lat, bnds\) float64 (3kB )?\.\.\.
+    time_bnds      \(time, bnds\) datetime64\[ns\] (80B )?\.\.\.
+Dimensions without coordinates: bnds
+Data variables:
+    precipitation  \(time, lat, lon\) float64 (3MB )?...
+    temperature    \(time, lat, lon\) float64 (3MB )?...
+    soil_moisture  \(time, lat, lon\) float64 (3MB )?...
+Attributes:
+    Conventions:           CF-1.7
+    title:                 Test Cube
+    time_coverage_start:   2010-01-01T00:00:00
+    time_coverage_end:     2010-01-06T00:00:00
+    geospatial_lon_min:    -180.0
+    geospatial_lon_max:    180.0
+    geospatial_lon_units:  degrees_east
+    geospatial_lat_min:    -90.0
+    geospatial_lat_max:    90.0
+    geospatial_lat_units:  degrees_north
+"""
 
-        possible_outputs = [
-            output_template % '2:00:00',  # for xarray v0.15.1 and below
-            output_template % '...',  # for xarray v0.16.0 and above
-            output_template_1 % '...'  # for xarray v2022.06.0 and above
-        ]
-        print(result.output)
-        self.assertIn(result.output, possible_outputs)
+        self.assertRegex(result.output, output_regex)
         self.assertEqual(0, result.exit_code)

--- a/test/core/test_dump.py
+++ b/test/core/test_dump.py
@@ -36,11 +36,13 @@ class DumpDatasetTest(unittest.TestCase):
         self.assertIn("    _FillValue:  999.0\n", text)
 
         text = dump_dataset(dataset, ["precipitation"])
-        self.assertIn("<xarray.DataArray 'precipitation' (time: 5, lat: 180, lon: 360)>\n", text)
+        precip_regex = r"<xarray\.DataArray 'precipitation' .*" \
+            r"\(time: 5, lat: 180, lon: 360\)>( Size: 3MB)?\n"
+        self.assertRegex(text, precip_regex)
         self.assertNotIn("Encoding:\n", text)
         self.assertNotIn("    _FillValue:  999.0", text)
 
         text = dump_dataset(dataset, ["precipitation"], show_var_encoding=True)
-        self.assertIn("<xarray.DataArray 'precipitation' (time: 5, lat: 180, lon: 360)>\n", text)
+        self.assertRegex(text, precip_regex)
         self.assertIn("Encoding:\n", text)
         self.assertIn("    _FillValue:  999.0", text)

--- a/xcube/core/new.py
+++ b/xcube/core/new.py
@@ -52,7 +52,9 @@ def new_cube(title='Test Cube',
              drop_bounds=False,
              variables=None,
              crs=None,
-             crs_name=None):
+             crs_name=None,
+             time_encoding_dtype="int64"
+             ):
     """
     Create a new empty cube. Useful for creating cubes templates with
     predefined coordinate variables and metadata. The function is also
@@ -102,6 +104,8 @@ def new_cube(title='Test Cube',
         of pyproj.CRS or None
     :param crs_name: Name of the variable that will
         hold the CRS information. Ignored, if *crs* is not given.
+    :param time_encoding_dtype: data type used to encode the time variable when
+        serializing the dataset
     :return: A cube instance
     """
     y_dtype = y_dtype if y_dtype is not None else y_dtype
@@ -163,6 +167,7 @@ def new_cube(title='Test Cube',
     time_var = xr.DataArray(time_data, dims=time_name)
     time_var.encoding['units'] = time_units
     time_var.encoding['calendar'] = time_calendar
+    time_var.encoding["dtype"] = time_encoding_dtype
 
     coords = {x_name: x_var, y_name: y_var, time_name: time_var}
     if not drop_bounds:
@@ -203,6 +208,7 @@ def new_cube(title='Test Cube',
                                      dims=(time_name, bnds_dim))
         time_bnds_var.encoding['units'] = time_units
         time_bnds_var.encoding['calendar'] = time_calendar
+        time_bnds_var.encoding["dtype"] = time_encoding_dtype
 
         time_var.attrs['bounds'] = time_bnds_name
 

--- a/xcube/core/normalize.py
+++ b/xcube/core/normalize.py
@@ -270,8 +270,8 @@ def _normalize_zonal_lat_lon(ds: xr.Dataset) -> xr.Dataset:
             var_chunk_sizes.append(lat_chunk_size)
             var_chunk_sizes.append(ds_zonal.lon.size)
     ds_zonal = ds_zonal.rename_dims({'latitude_centers': 'lat'})
-    ds_zonal = ds_zonal.assign_coords(lat=ds.latitude_centers.values)
     ds_zonal = ds_zonal.drop_vars('latitude_centers')
+    ds_zonal = ds_zonal.assign_coords(lat=ds.latitude_centers.values)
     ds_zonal = ds_zonal.transpose(..., 'lat', 'lon')
 
     has_lon_bnds = 'lon_bnds' in ds_zonal.coords or 'lon_bnds' in ds_zonal

